Src/Support/Google.Apis.Auth/OAuth2/ServiceAccountCredential.cs

```diff
@@ -221,29 +221,89 @@ public static ServiceAccountCredential FromServiceAccountData(Stream credentialD
         {
             if (!HasScopes && authUri != null)
             {
-                // TODO(jtattermusch): support caching of JWT access tokens per authUri, currently a new 
-                // JWT access token is created each time, which can hurt performance.
-                return CreateJwtAccessToken(authUri);
+                return await GetOrCreateJwtAccessTokenAsync(authUri).ConfigureAwait(false);
             }
             return await base.GetAccessTokenForRequestAsync(authUri, cancellationToken).ConfigureAwait(false);
         }
-    
+
+        private class JwtCacheEntry
+        {
+            public JwtCacheEntry(Task<string> jwtTask, string uri, DateTime expiryUtc)
+            {
+                JwtTask = jwtTask;
+                Uri = uri;
+                ExpiryUtc = expiryUtc;
+            }
+
+            public Task<string> JwtTask { get; }
+            public string Uri { get; }
+            public DateTime ExpiryUtc { get; }
+        }
+
+        // Internal for testing.
+        internal static readonly TimeSpan JwtLifetime = TimeSpan.FromMinutes(60);
+        internal static readonly TimeSpan JwtCacheExpiryWindow = TimeSpan.FromMinutes(5);
+        internal const int JwtCacheMaxSize = 100;
+        private readonly object _jwtLock = new object();
+        private LinkedList<JwtCacheEntry> _jwts = null;
+        private Dictionary<string, LinkedListNode<JwtCacheEntry>> _jwtCache = null;
+
+        private Task<string> GetOrCreateJwtAccessTokenAsync(string authUri)
+        {
+            var nowUtc = Clock.UtcNow;
+            lock (_jwtLock)
+            {
+                if (_jwtCache == null)
+                {
+                    // Create cache on demand, as many service credentials won't ever need one.
+                    _jwtCache = new Dictionary<string, LinkedListNode<JwtCacheEntry>>();
+                    _jwts = new LinkedList<JwtCacheEntry>();
+                }
+                if (_jwtCache.TryGetValue(authUri, out var cachedJwtNode))
+                {
+                    var jwtEntry = cachedJwtNode.Value;
+                    if (jwtEntry.ExpiryUtc - JwtCacheExpiryWindow > nowUtc)
+                    {
+                        // Cached JWT not expired, return it.
+                        return jwtEntry.JwtTask;
+                    }
+                    // Cached JWT is expired; remove it.
+                    _jwtCache.Remove(authUri);
+                    _jwts.Remove(cachedJwtNode);
+                }
+                // Create a new JWT.
+                var expiryUtc = nowUtc + JwtLifetime;
+                var jwtTask = Task.Run(() => CreateJwtAccessToken(authUri, nowUtc, expiryUtc));
```

Comment:
Out of interest, have we checked that the compute cost of creating a JWT is significantly greater than the compute cost of scheduling a task? I'd expect so, but...

Reply:
No I have not checked this. But I would be extremely surprised if scheduling a task takes more CPU time.